File: python/benchmark.py
import argparse
import threading
import time
from typing import Union

import influxdb
from influxdb_client import InfluxDBClient
from influxdb_client.client.write_api import WriteOptions


class Writer:
    def __init__(self, influxdb_client: Union[InfluxDBClient, influxdb.InfluxDBClient], batch_size: int,
                 flush_interval: int):
        self.client = influxdb_client

        if isinstance(influxdb_client, InfluxDBClient):
            self.write_api = self.client.write_api(WriteOptions(batch_size=batch_size, flush_interval=flush_interval))

    def write(self, id: int, measurement_name: str, iteration: int):
        pass


class WriterV1(Writer):
    def write(self, id: int, measurement_name: str, iteration: int, batch_size=0, flush_interval=0):
        line = "%s,id=%s temperature=%d %d" % (measurement_name, id, time.time(), iteration)
        self.client.write_points(points=line, protocol="line")

    def countRows(self):
        pass


class WriterV2(Writer):
    def write(self, id: int, measurement_name: str, iteration: int):
        line = "%s,id=%s temperature=%d %d" % (measurement_name, id, time.time(), iteration)
        self.write_api.write(bucket="my-bucket", org="my-org", record=line)

    def countRows(self, measurement_name):
        query = '''from(bucket:"my-bucket")
        |> range(start: 0, stop: now())
        |> filter(fn: (r) => r._measurement == "{}")
        |> pivot(rowKey: ["_time"], columnKey: ["_field"], valueColumn: "_value")
        |> drop(columns: ["id", "host"]) |> count(column: "temperature")'''.format(measurement_name)

        result = self.client.query_api().query(query=query)
        if (len(result) > 0):
            return result[0].records[0]['temperature']
        return 0


def worker(stop_event, id: int, measurement_name, seconds_count, line_protocol_count):
    start_time = time.time()
    for i in range(1, seconds_count + 1):
        iteration_time_start = time.time()
        if stop_event.is_set():
            return

        if id == 0:
            print("writing iterations:: {}/{} id:{}".format(i, seconds_count, id))

        start = i * line_protocol_count
        stop = start + line_protocol_count

        for j in range(start, stop):

            if time.time() > start_time + seconds_count:
                print(f"Time elapsed {id}", id)
                return

            if stop_event.is_set():
                return
            writer.write(id, measurement_name, j)

        if stop_event.is_set():
            return

        iteration_duration = time.time() - iteration_time_start
        print(f"iteration duration {iteration_duration:.2f}s")

        if iteration_duration < 1:
            time.sleep(1 - iteration_duration)


def main():
    global writer

    parser = argparse.ArgumentParser(description='InfluxDB 2.0 client lib benchmark')
    parser.add_argument("-type", default="CLIENT_PYTHON_V2",
                        help="Type of writer (default 'CLIENT_PYTHON_V2'; CLIENT_PYTHON_V1, CLIENT_PYTHON_V2)")
    # 2000
    parser.add_argument("-threadsCount", type=int, default=200, help="how much Thread use to write into InfluxDB")
    parser.add_argument("-secondsCount", type=int, default=60, help="how long write into InfluxDB")
    parser.add_argument("-batchSize", type=int, default=50000, help="batch size")
    parser.add_argument("-flushInterval", type=int, default=10000, help="buffer flush interval")
    parser.add_argument("-lineProtocolsCount", type=int, default=10, help="how much data writes in one batch")
    parser.add_argument("-skipCount", action="store_true", help="skip query for counting rows on end of benchmark")
    parser.add_argument("-measurementName", default="sensor_%d" % (time.time()), help="measurement name")

    args = parser.parse_args()

    writer_type = args.type
    threads_count = args.threadsCount
    seconds_count = args.secondsCount
    batch_size = args.batchSize
    flush_interval = args.flushInterval
    line_protocols_count = args.lineProtocolsCount
    skip_count = args.skipCount
    measurement_name = args.measurementName

    expected = threads_count * seconds_count * line_protocols_count

    print()
    print("------------- %s -------------" % writer_type)
    print()
    print()
    print("measurement:        %s " % measurement_name)
    print("threadsCount:       %s " % threads_count)
    print("secondsCount:       %s " % seconds_count)
    print("lineProtocolsCount: %s " % line_protocols_count)
    print()
    print("batchSize:          %s " % batch_size)
    print("flushInterval:      %s " % flush_interval)
    print()
    print("expected size: %s" % expected)
    print()
    print(args.type)

    if writer_type == 'CLIENT_PYTHON_V2':
        writer = WriterV2(
            influxdb_client=InfluxDBClient(url="http://localhost:9999", token="my-token", org="my-org", debug=False),
            batch_size=batch_size,
            flush_interval=flush_interval)
    if writer_type == 'CLIENT_PYTHON_V1':
        writer = WriterV1(influxdb.InfluxDBClient('localhost', 8086, 'root', 'root', 'iot_writes'))

    stop_event = threading.Event()

    if threads_count > 1:
        threads = []
        for i in range(threads_count):
            # print("Prepare thread %d " % i)
            t = threading.Thread(target=worker, daemon=True,
                                 args=(stop_event, i, measurement_name, seconds_count, line_protocols_count))
            threads.append(t)

        for thread in threads:
            thread.start()

        # sleep main thread
        time.sleep(seconds_count)

        # send stop event to all threads
        print("Stop event!")
        stop_event.set()

        # wait threads to finish
        for thread in threads:
            thread.join(timeout=seconds_count)

    else:
        worker(stop_event, 0, measurement_name, seconds_count, line_protocols_count)

    if not skip_count:
        count = writer.countRows(measurement_name=measurement_name)

        print()
        print("Results:")
        print(f"-> expected:        {expected} ")
        print(f"-> total:           {count} ")
        print(f"-> rate [%]:        {(count / expected) * 100} ")
        print(f"-> rate [msg/sec]:  {count / seconds_count} ")

        print("Written records {}:{}".format(writer_type, count))


if __name__ == '__main__':
    main()

File: ruby/benchmark.rb
require 'optparse'
require 'ostruct'
require 'influxdb'
require 'influxdb2/client'

# Ruby benchmark
module Benchmark
  # Parser arguments
  class Parser
    def self.parse(args)
      options = OpenStruct.new
      options.type = 'CLIENT_RUBY_V1'
      options.threads_count = 200
      options.seconds_count = 60
      options.batch_size = 50_000
      options.flush_interval = 10_000
      options.line_protocols_count = 10
      options.skip_count = false
      options.measurement_name = "sensor_#{Time.new.to_i}"
      types = %w[CLIENT_RUBY_V1 CLIENT_RUBY_V2]

      opt_parser = OptionParser.new do |opts|
        opts.on('--type VALUE', types,
                'Type of writer (default "CLIENT_RUBY_V1"; CLIENT_RUBY_V1, CLIENT_RUBY_V2)') do |value|
          options.type = value
        end

        opts.on('--threadsCount VALUE', Integer, 'how much Thread use to write into InfluxDB') do |value|
          options.threads_count = value
        end

        opts.on('--secondsCount VALUE', Integer, 'how long write into InfluxDB') do |value|
          options.seconds_count = value
        end

        opts.on('--batchSize VALUE', Integer, 'batch size') do |value|
          options.batch_size = value
        end

        opts.on('--flushInterval VALUE', Integer, 'buffer flush interval') do |value|
          options.flush_interval = value
        end

        opts.on('--lineProtocolsCount VALUE', Integer, 'how much data writes in one batch') do |value|
          options.line_protocols_count = value
        end

        opts.on('--skipCount', 'skip query for counting rows on end of benchmark') do
          options.skip_count = true
        end

        opts.on('--measurementName VALUE', 'measurement name') do |value|
          options.measurement_name = value
        end

        opts.on('-?', '--help', 'Prints this help') do
          puts opts
        end
      end

      begin
        opt_parser.parse(args)
      rescue StandardError => e
        puts "Exception encountered: #{e}"
        exit 1
      end

      options
    end
  end

  # default writer
  #
  class Writer
    def initialize(influxdb_client)
      @client = influxdb_client
    end

    attr_reader :client
  end

  # Writer for InfluxDB V1
  #
  class WriterV1 < Writer
    def initialize(influxdb_client)
      super(influxdb_client)
    end

    def write(id, measurement_name:, iteration:)
      data = {id: id, values: {temprerature: rand(-10..35)}, timestamp: iteration}
      @client.write_point(measurement_name, data)
    end

    def count_rows(measurement_name)
      query = "select count(*) from #{measurement_name}"
      result = @client.query(query)
      result.empty? ? 0 : result[0]['values'][0]['count_temprerature']
    end
  end

  # Writer for InfluxDB V2
  #
  class WriterV2 < Writer
    def initialize(influxdb_client, batch_size:, flush_interval:)
      super(influxdb_client)

      @write_api = @client.create_write_api(
        write_options: InfluxDB2::WriteOptions.new(
          write_type: InfluxDB2::WriteType::BATCHING,
          batch_size: batch_size,
          flush_interval: flush_interval
        )
      )
    end

    def write(id, measurement_name:, iteration:)
      line = "#{measurement_name},id=#{id} temperature=#{rand(-10..35)} #{iteration}"
      @write_api.write(bucket: 'my-bucket', org: 'my-org', data: line)
    end

    def count_rows(measurement_name)
      query = "from(bucket:\"my-bucket\") \n
        |> range(start: 0, stop: now()) \n
        |> filter(fn: (r) => r._measurement == \"#{measurement_name}\") \n
        |> pivot(rowKey: [\"_time\"], columnKey: [\"_field\"],valueColumn: \"_value\") \n
        |> drop(columns: [\"id\", \"host\"]) \n
        |> count(column: \"temperature\")"

      result = @client.create_query_api.query(query: query, org: 'my-org')
      result.empty? ? 0 : result[0].records[0].values['temperature']
    end
  end

  # Worker which manages writing threads
  #
  class Worker
    attr_reader :threads

    def initialize(writer, options)
      @writer = writer
      @stop = false
      @options = options

      @threads = []
      @options.threads_count.times do |id|
        @threads << Thread.new do
          _work(id)
        end
      end
    end

    def stop!
      @stop = true
    end

    private

    def _work(id)
      start_time = Time.new.to_f

      (1..@options.seconds_count + 1).each do |i|
        iteration_time_start = Time.new.to_f

        break if Time.new.to_f > start_time + @options.seconds_count || @stop

        puts "writing iterations:: #{i}/#{@options.seconds_count}" if id == 0

        start = i * @options.line_protocols_count
        stop = start + @options.line_protocols_count

        (start..stop).each do |j|
          if Time.new.to_f > start_time + @options.seconds_count
            # puts "#{id} Time elapsed"
            break
          end

          break if Time.new.to_f > start_time + @options.seconds_count || @stop

          @writer.write(id, measurement_name: @options.measurement_name, iteration: j)

          break if @stop

          iteration_duration = Time.new.to_f - iteration_time_start.to_f
          # puts "iteration duration #{iteration_duration}s"

          sleep(1 - iteration_duration) if iteration_duration < 1
        end
      end
    end
  end

  class Client2 < InfluxDB2::Client
    def close!
      # do not flush queue
    end
  end

  options = Parser.parse(ARGV)

  expected = options.threads_count * options.seconds_count *
             options.line_protocols_count

  puts "\n"
  print(format("------------- #{options.type} -------------\n"))
  puts "\n"
  print("measurement:        #{options.measurement_name} \n")
  print("threadsCount:       #{options.threads_count} \n")
  print("secondsCount:       #{options.seconds_count} \n")
  print("lineProtocolsCount: #{options.line_protocols_count} \n")
  puts "\n"
  print("batchSize:          #{options.batch_size} \n")
  print("flushInterval:      #{options.flush_interval} \n")
  puts "\n"
  puts format("expected size:      #{expected} \n")
  puts "\n"

  writer = nil

  if options.type == 'CLIENT_RUBY_V2'
    writer = WriterV2.new(Client2.new('http://localhost:9999',
                                                'my-token',
                                                org: 'my-org',
                                                bucket: 'my-bucket',
                                                precision: InfluxDB2::WritePrecision::NANOSECOND,
                                                use_ssl: false),
                          batch_size: options.batch_size,
                          flush_interval: options.flush_interval)
  elsif options.type == 'CLIENT_RUBY_V1'
    async_options = {max_post_points: options.batch_size}
    writer = WriterV1.new(InfluxDB::Client.new(url: 'http://localhost:8086',
                                               database: 'iot_writes',
                                               user: 'root',
                                               password: 'root',
                                               async: async_options))
  end

  unless writer.nil?
    worker = Worker.new(writer, options)

    # sleep main thread
    sleep(options.seconds_count)

    # stop all threads
    puts "Stoping threads!\n"
    worker.stop!

    worker.threads.each do |t|
      t.join
    end

    puts "All threads finished"

    unless options.skip_count
      count = writer.count_rows(options.measurement_name)

      puts "\n"
      puts "Results:\n"
      puts "-> expected:        #{expected} \n"
      puts "-> total:           #{count} \n"
      puts "-> rate [%]:        #{(count.to_f / expected.to_f) * 100} \n"
      puts "-> rate [msg/sec]:  #{count / options.seconds_count} \n"

      puts "Written records #{options.type}:#{count}"
    end
  end
end
